jaclang_jaseci/plugins/common.py

```diff
@@ -0,0 +1,1174 @@
+
+    def __deepcopy__(self, memo: dict) -> "DocAnchor[DA]":
+        """Override deepcopy implementation."""
+        memo[id(self)] = self
+        return self
+
+
+class DocArchitype(Generic[DA]):
+    """DocAnchor Class Handler."""
+
+    _jac_type_: JType
+
+    class Collection(ArchCollection):
+        """Default DocArchitype Collection."""
+
+        pass
+
+    @property
+    def _jac_doc_(self) -> DocAnchor[DA]:
+        if isinstance(jd := getattr(self, "__jac_doc__", None), DocAnchor):
+            return jd
+
+        jctx: JacContext = JCONTEXT.get()
+        jd = self.__jac_doc__ = DocAnchor[DA](
+            type=self._jac_type_,
+            name=self.__class__.__name__,
+            root=jctx.get_root_id(),
+        )
+        if isinstance(self, DocArchitype):
+            jd.arch = self
+
+        return jd
+
+    @_jac_doc_.setter
+    def _jac_doc_(self, val: DocAnchor[DA]) -> None:
+        self.__jac_doc__ = val
+
+    def get_context(self) -> dict:
+        """Retrieve context dictionary."""
+        ctx = {}
+        if is_dataclass(self) and not isinstance(self, type):
+            ctx = asdict(self)
+
+        return {"context": ctx}
+
+    async def propagate_save(
+        self, changes: dict, session: AsyncIOMotorClientSession
+    ) -> list[InsertOne[Any] | DeleteMany | DeleteOne | UpdateMany | UpdateOne]:
+        """Propagate saving."""
+        _ops: list[tuple[dict[str, ObjectId], dict[str, Any]]] = []
+
+        jd_id = self._jac_doc_.id
+
+        for ops in [
+            ("$pull", "$in"),
+            ("$addToSet", "$each"),
+        ]:
+            _list = None
+            target = changes.get(ops[0], {})
+            for op in target.values():
+                if _set := op[ops[1]]:
+                    _list = op[ops[1]] = list(_set)
+                    for idx, danch in enumerate(_list):
+                        if isinstance(danch, DocAnchor):
+                            if not danch.connected and danch.arch:
+                                await danch.arch.save(session)
+                            _list[idx] = danch.ref_id
+            if _list:
+                _ops.append(({"_id": jd_id}, {ops[0]: target}))
+
+        if _set := changes.get("$set"):
+            if _ops:
+                _ops[0][1].update({"$set": _set})
+            else:
+                _ops.append(({"_id": jd_id}, {"$set": _set}))
+
+        return [UpdateOne(*_op) for _op in _ops]
+
+    async def save(
+        self, session: Optional[AsyncIOMotorClientSession] = None
+    ) -> DocAnchor[DA]:
+        """Upsert Architype."""
+        raise Exception("Save function not implemented yet!")
+
+    async def save_with_session(self) -> None:
+        """Upsert Architype with session."""
+        async with await ArchCollection.get_session() as session:
+            async with session.start_transaction():
+                try:
+                    await self.save(session)
+                    await session.commit_transaction()
```

Comment:
If i understand this correctly, all nodes will be saved in one commit and all edges will be saved in a separate commit. is that correct?

Reply:
And any delete will be in another commit, handled by `destroy`?